src/main/java/dk/sdu/mmmi/modulemon/MCTSBattleAI/MCTSBattleAI.java

```diff
@@ -0,0 +1,268 @@
+            if(uct > bestUCT){
+                bestChild = child;
+                bestUCT = uct;
+            }
+        }
+        return bestChild;
+    }
+
+    private float calculateUCT(Node child, float exploration_coefficient) {
+        var exploitation = child.getReward() / child.getTimesVisited();
+        var exploration = Math.sqrt((2*Math.log(child.getParent().getTimesVisited()))/child.getTimesVisited());
+        return (float) (exploitation + exploration_coefficient * exploration);
+    }
+
+    private Node expandNode(Node node) {
+        Object action = untriedAction(node);
+        if(action instanceof Integer a && a == -1){
+            throw new IllegalStateException("No possible action was found when expanding node");
+        }
+
+        Node child;
+        if(action instanceof IMonsterMove move){
+            child = new Node(
+                    battleSimulation.simulateDoMove(participantToControl, move, battleSimulation.getState()),
+                    node,
+                    move);
+        } else if(action instanceof IMonster monster){
+            child = new Node(
+                    battleSimulation.simulateSwitchMonster(participantToControl, monster, battleSimulation.getState()),
+                    node,
+                    monster);
+        } else {
+            throw new IllegalStateException("No action was chosen when expanding node");
+        }
+
+        return child;
+    }
+
+    private Object untriedAction(Node node) {
+        var possibleMoves = participantToControl.getActiveMonster().getMoves();
+        var possibleSwitchActions = participantToControl.getMonsterTeam().stream()
+                .filter(m -> m.getHitPoints() > 0 && m != participantToControl.getActiveMonster()).toList();
+        List<Object> possibleActions = new ArrayList<Object>();
+        for (IMonsterMove move : possibleMoves) {
+            if(node.getChildren().stream().noneMatch(c -> c.getParentMove() == move)){
+                possibleActions.add(move);
+            }
+        }
+        for (IMonster monster : possibleSwitchActions) {
+            if(node.getChildren().stream().noneMatch(c -> c.getParentSwitch() == monster)){
+                possibleActions.add(monster);
+            }
+        }
+        var possibleActionCount = possibleActions.size();
+        if(possibleActionCount == 0) return -1;
+        else{
+            var rand = new Random();
+            return possibleActions.get(rand.nextInt((int) possibleActionCount));
+        }
+    }
+
+    private boolean fullyExpanded(Node node) {
+        var moveCount = participantToControl.getActiveMonster().getMoves().size();
+        var switchCount = participantToControl.getMonsterTeam().stream()
+                .filter(m -> m.getHitPoints() > 0 && m != participantToControl.getActiveMonster())
+                .count();
+        return node.getChildren().size() == (moveCount + switchCount);
```

Comment:
```suggestion
        return node.getChildren().size() >= (moveCount + switchCount);
```

Seems like a bug waiting to happen.